zend_abstract_interface/config/config_decode.c

```diff
@@ -0,0 +1,201 @@
+     * character used in the conversion is stored in the location referenced
+     * by endptr. If no conversion is performed, zero is returned and the value
+     * of nptr is stored in the location referenced by endptr.
+     */
+    if (endptr != value.ptr && errno == 0) {
+        ZVAL_DOUBLE(decoded_value, result);
+        return true;
+    }
+    return false;
+}
+
+/* As with strtod(), strtoll() also supports conversion of constants and
+ * ill-formatted numbers so we add a number-format check.
+ */
+static bool zai_config_is_valid_int_format(const char *str) {
+    while (*str) {
+        if (!isdigit(*str) && !isspace(*str)) return false;
+        str++;
+    }
+    return true;
+}
+
+static bool zai_config_decode_int(zai_string_view value, zval *decoded_value) {
+    if (!zai_config_is_valid_int_format(value.ptr)) return false;
+
+    char *endptr = (char *)value.ptr;
+
+    errno = 0;
+    long long int result = strtoll(value.ptr, &endptr, 10);
+
+    if (endptr != value.ptr && errno == 0) {
+        ZVAL_LONG(decoded_value, result);
+        return true;
+    }
+    return false;
+}
+
+static void zai_config_pstr_dtor(zval *zval_ptr) {
+    ZEND_ASSERT(Z_TYPE_P(zval_ptr) == IS_STRING);
+    zend_string_release_ex(Z_STR_P(zval_ptr), 1);
+}
+
+static bool zai_config_decode_map(zai_string_view value, zval *decoded_value, bool persistent) {
+    zval tmp;
+    if (persistent) {
+        HashTable *ht = malloc(sizeof(HashTable));
+        zend_hash_init(ht, 8, NULL, zai_config_pstr_dtor, /* persistent */ 1);
+        ZVAL_ARR(&tmp, ht);
+    } else {
+        array_init(&tmp);
+    }
+
+    // TODO Extract this
+    char *data = (char *)value.ptr;
+    if (data && *data) {  // non-empty
+        const char *key_start, *key_end, *value_start, *value_end;
+        do {
+            if (*data != ',' && *data != ' ' && *data != '\t' && *data != '\n') {
+                key_start = key_end = data;
+                while (*++data) {
+                    if (*data == ':') {
+                        while (*++data && (*data == ' ' || *data == '\t' || *data == '\n'))
+                            ;
+
+                        if (!*data) {
+                            break;
+                        }
+
+                        value_start = value_end = data;
+                        if (*data == ',') {
+                            --value_end;  // empty string instead of single char
+                        } else {
+                            while (*++data && *data != ',') {
+                                if (*data != ' ' && *data != '\t' && *data != '\n') {
+                                    value_end = data;
+                                }
+                            }
+                        }
+
+                        zval val;
+                        zend_string *key = zend_string_init(key_start, key_end - key_start + 1, persistent);
+                        if (persistent) {
```

Comment:
`ZVAL_NEW_STR(&val, zend_string_init(value_start, value_end - value_start + 1, persistent));` maybe instead of branching here and duplicating args

Reply:
Similarly in zai_config_decode_string below

Reply:
Nice catch. 👍